src/laplace.jl

```diff
@@ -0,0 +1,332 @@
+        lfx = lf(xs)
+        ignore_ad() do
+            # Zygote does not like the try/catch within @info etc.
+            @debug "Hyperparameters: $theta"
+            if initialize_f
+                f .= mean(lfx.fx)
+            end
+        end
+        f_opt, lml = laplace_f_and_lml(
+            lfx, ys; f_init=f, maxiter=newton_maxiter, callback=newton_callback
+        )
+        ignore_ad() do
+            if newton_warmstart
+                f .= f_opt
+                initialize_f = false
+            end
+        end
+        return -lml
+    end
+
+    return objective
+end
+
+function build_laplace_objective(build_latent_gp, xs, ys; kwargs...)
+    f = similar(xs, length(xs))  # will be mutated in-place to "warm-start" the Newton steps
+    return build_laplace_objective!(f, build_latent_gp, xs, ys; kwargs...)
+end
+
+function laplace_f_cov(cache)
+    # (K⁻¹ + W)⁻¹
+    # = (√W⁻¹) (√W⁻¹ (K⁻¹ + W) √W⁻¹)⁻¹ (√W⁻¹)
+    # = (√W⁻¹) (√W⁻¹ K⁻¹ √W⁻¹ + I)⁻¹ (√W⁻¹)
+    # ; (I + C⁻¹)⁻¹ = I - (I + C)⁻¹
+    # = (√W⁻¹) (I - (I + √W K √W)⁻¹) (√W⁻¹)
+    # = (√W⁻¹) (I - B⁻¹) (√W⁻¹)
+    B_ch = cache.B_ch
+    Wsqrt_inv = inv(cache.Wsqrt)
+    return Wsqrt_inv * (I - inv(B_ch)) * Wsqrt_inv
+end
+
+function LaplaceResult(fnew, cache)
+    f = cache.f
+    f_cov = laplace_f_cov(cache)
+    q = MvNormal(f, AbstractGPs._symmetric(f_cov))
+    lml_approx = _laplace_lml(f, cache)
+
+    return (; fnew, f_cov, q, lml_approx, cache...)
+end
+
+"""
+    laplace_steps(lfx::LatentFiniteGP, ys; newton_kwargs...)
+
+For demonstration purposes: returns an array of all the intermediate
+approximations of each Newton step.
+
+If you are only interested in the actual posterior, use
+`posterior(::LaplaceApproximation, ...`.
+
+TODO figure out how to get the `@ref` to work...
+"""
+function laplace_steps(lfx::LatentFiniteGP, ys; newton_kwargs...)
+    dist_y_given_f, K, newton_kwargs = _check_laplace_inputs(lfx, ys; newton_kwargs...)
+
+    res_array = []
+
+    function store_result!(fnew, cache)
+        return push!(res_array, LaplaceResult(fnew, cache))
+    end
+
+    _ = newton_inner_loop(dist_y_given_f, ys, K; newton_kwargs..., callback=store_result!)
+
+    return res_array
+end
+
+struct LaplaceApproximation{Tkw}
+    newton_kwargs::Tkw
+end
+
+LaplaceApproximation(; newton_kwargs...) = LaplaceApproximation((; newton_kwargs...))
+
+function approx_lml(la::LaplaceApproximation, lfx::LatentFiniteGP, ys)
+    return laplace_lml(lfx, ys; la.newton_kwargs...)
+end
+
+function AbstractGPs.posterior(la::LaplaceApproximation, lfx::LatentFiniteGP, ys)
+    dist_y_given_f, K, newton_kwargs = _check_laplace_inputs(lfx, ys; la.newton_kwargs...)
+    _, cache = _newton_inner_loop(dist_y_given_f, ys, K; newton_kwargs...)
+    # TODO: should we run newton_inner_loop() and _laplace_train_intermediates() explicitly?
+    f_post = ApproxPosteriorGP(la, lfx.fx, cache)
+    # TODO: instead of lfx.fx, should we store lfx itself (including lik)?
```

Comment:
> this is a more general question for the LatentGP API

Seems plausible. An issue to discuss further would be good.

Reply:
https://github.com/JuliaGaussianProcesses/AbstractGPs.jl/issues/223